index.js

```diff
@@ -29,75 +28,35 @@ module.exports = function(moduleName, options) {
   var stopDir = options.stopDir || oshomedir();
   var splitSearchPath = splitPath(options.cwd);
 
-  return new Promise(function(resolve, reject) {
-
-    find();
-
-    function find() {
-      if (options.configPath) {
-        loadDefinedFile(options.configPath, options.format).then(function(result) {
-          return finishWith(result);
-        }).catch(reject);
-        return;
-      }
-
-      var currentSearchPath = joinPath(splitSearchPath);
-
-      var search = Promise.resolve(null);
-
-      if (options.packageProp) {
-        search = search.then(function(result) {
-          if (result) return finishWith(result);
-          return loadPackageProp(currentSearchPath, options.packageProp);
-        });
-      }
-
-      if (options.rc) {
-        search = search.then(function(result) {
-          if (result) return finishWith(result);
-          return loadRc(path.join(currentSearchPath, options.rc), {
-            strictJson: options.rcStrictJson,
-          });
-        });
-      }
-
-      if (options.js) {
-        search = search.then(function(result) {
-          if (result) return finishWith(result);
-          return loadJs(path.join(currentSearchPath, options.js));
-        });
-      }
-
-      search.then(function(result) {
-        if (result) return finishWith(result);
-        return moveUpOrGiveUp(currentSearchPath);
-      })
-        .then(function(result) {
-          if (result === DONE) return resolve(null);
-          find();
-        })
-        .catch(function(err) {
-          if (err === DONE) return;
-          reject(err);
-        });
+  if (options.configPath) {
+    return loadDefinedFile(options.configPath, options.format);
+  }
 
-      function moveUpOrGiveUp(searchPath) {
-        return new Promise(function(resolve) {
-          // Notice the mutation of splitSearchPath
-          if (searchPath === stopDir || !splitSearchPath.pop()) {
-            resolve(DONE);
-          } else {
-            resolve(null);
-          }
-        });
+  function find() {
+    var currentSearchPath = joinPath(splitSearchPath);
+
+    return Promise.resolve(null).then(function() {
+      if (!options.packageProp) return;
+      return loadPackageProp(currentSearchPath, options.packageProp);
+    }).then(function(result) {
+      if (result || !options.rc) return result;
+      return loadRc(path.join(currentSearchPath, options.rc), {
+        strictJson: options.rcStrictJson,
+      });
```

Comment:
Oooo, this is cool: taking advantage of the fact that you can return a synchronous value *or* a Promise and they are treated the same way. Working on this module has taught me a lot about Promises :)